src/libutil/git.cc

```diff
@@ -79,47 +77,91 @@ void parse(
             checkInterrupt();
             buf.resize(std::min((unsigned long long)buf.capacity(), left));
             source(buf);
-            sink.receiveContents(buf);
+            crf(buf);
             left -= buf.size();
         }
-    } else if (type == "tree ") {
-        unsigned long long size = std::stoi(getStringUntil(source, 0));
-        unsigned long long left = size;
+    });
+}
 
-        sink.createDirectory(sinkPath);
+void parseTree(
+    ParseSink & sink,
+    const Path & sinkPath,
+    Source & source,
+    std::function<SinkHook> hook,
+    const ExperimentalFeatureSettings & xpSettings)
+{
+    unsigned long long size = std::stoi(getStringUntil(source, 0));
+    unsigned long long left = size;
 
-        while (left) {
-            std::string perms = getStringUntil(source, ' ');
-            left -= perms.size();
-            left -= 1;
+    sink.createDirectory(sinkPath);
 
-            RawMode rawMode = std::stoi(perms, 0, 8);
-            auto modeOpt = decodeMode(rawMode);
-            if (!modeOpt)
-                throw Error("Unknown Git permission: %o", perms);
-            auto mode = std::move(*modeOpt);
+    while (left) {
+        std::string perms = getStringUntil(source, ' ');
+        left -= perms.size();
+        left -= 1;
 
-            std::string name = getStringUntil(source, '\0');
-            left -= name.size();
-            left -= 1;
+        RawMode rawMode = std::stoi(perms, 0, 8);
+        auto modeOpt = decodeMode(rawMode);
+        if (!modeOpt)
+            throw Error("Unknown Git permission: %o", perms);
+        auto mode = std::move(*modeOpt);
 
-            std::string hashs = getString(source, 20);
-            left -= 20;
+        std::string name = getStringUntil(source, '\0');
+        left -= name.size();
+        left -= 1;
 
-            Hash hash(HashAlgorithm::SHA1);
-            std::copy(hashs.begin(), hashs.end(), hash.hash);
+        std::string hashs = getString(source, 20);
+        left -= 20;
 
-            hook(name, TreeEntry {
-                .mode = mode,
-                .hash = hash,
-            });
+        Hash hash(HashAlgorithm::SHA1);
+        std::copy(hashs.begin(), hashs.end(), hash.hash);
 
-            if (mode == Mode::Executable)
-                sink.isExecutable();
```

Comment:
This was a bad hack that no longer works. The changes to this file are to make a better solution instead.

Reply:
This file will be easier to review with whitespace disabled